extensions/amp-script/0.1/amp-script.js

```diff
@@ -210,34 +225,94 @@ export class AmpScript extends AMP.BaseElement {
+   *
+   * Returns promise that resolves with the script contents or rejected if the
+   * fetch fails or if the script fails CORS checks.
+   *
+   * Returns null if script reference is missing.
+   *
+   * @param {string} debugId An element identifier for error messages.
    * @return {?Promise<string>}
    * @private
    */
-  getAuthorScript_() {
+  getAuthorScript_(debugId) {
     const authorUrl = this.element.getAttribute('src');
     if (authorUrl) {
-      return Services.xhrFor(this.win)
-        .fetchText(authorUrl, {ampCors: false})
-        .then(r => r.text());
+      const urlService = Services.urlForDoc(this.element);
+
+      const docOrigin = urlService.getSourceOrigin(this.getAmpDoc().getUrl());
+      const scriptOrigin = urlService.parse(authorUrl).origin;
+      const sameOrigin = docOrigin === scriptOrigin;
+
+      return this.fetchAuthorScript_(authorUrl, sameOrigin, debugId);
     } else {
       const id = this.element.getAttribute('script');
       if (id) {
         const local = this.getAmpDoc().getElementById(id);
+        userAssert(
+          local,
+          '[%s] %s could not find element with #%s.',
+          TAG,
+          debugId,
+          id
+        );
         const target = local.getAttribute('target');
         userAssert(
           target === 'amp-script',
           '[%s] script#%s must have target="amp-script".',
           TAG,
           id
         );
-        return Promise.resolve(local.textContent);
+        const script = local.textContent;
+        return this.service_.checkSha384(script, debugId).then(() => script);
       }
     }
+    // No [src] or [script].
     return null;
   }
 
+  /**
+   * @param {string} authorUrl
+   * @param {boolean} sameOrigin Does `authorUrl` have same origin as the page?
+   * @param {string} debugId An element identifier for error messages.
+   * @return {!Promise<string>}
+   */
+  fetchAuthorScript_(authorUrl, sameOrigin, debugId) {
+    const init = {ampCors: false};
+    // Disallow redirects for same-origin scripts.
+    if (sameOrigin) {
+      // TODO(choumx): Polyfill this in our fetch polyfill.
+      init.redirect = 'error';
+    }
+    return Services.xhrFor(this.win)
+      .fetchText(authorUrl, init)
+      .then(r => {
+        if (sameOrigin) {
+          // Disallow non-JS content type for same-origin scripts.
+          const contentType = r.headers.get('Content-Type');
+          if (
+            !contentType ||
+            !startsWith(contentType, 'application/javascript')
+          ) {
+            throw user().createError(
+              '[%s] %s has Content-Type: "%s". ' +
+                'Only "application/javascript" is allowed.',
+              TAG,
+              debugId,
+              contentType
+            );
+          }
+          return r.text();
+        } else {
+          // For cross-origin, verify hash of script itself.
+          return r.text().then(text => {
```

Comment:
We could verify the content-type, too.

Reply:
Don't think it's necessary since we're already doing the safer check.